plugins/inputs/diskio/diskio_linux.go

```diff
@@ -37,9 +35,31 @@ func (s *DiskIO) diskInfo(devName string) (map[string]string, error) {
 		return ic.values, nil
 	}
 
-	major := unix.Major(uint64(stat.Rdev))
-	minor := unix.Minor(uint64(stat.Rdev))
-	udevDataPath := fmt.Sprintf("%s/b%d:%d", udevPath, major, minor)
+	var udevDataPath string
+	if ok && len(ic.udevDataPath) > 0 {
+		// We can reuse the udev data path from a "previous" entry.
+		// This allows us to also "poison" it during test scenarios
+		udevDataPath = ic.udevDataPath
+	} else {
+		major := unix.Major(uint64(stat.Rdev))
+		minor := unix.Minor(uint64(stat.Rdev))
+		udevDataPath = fmt.Sprintf("/run/udev/data/b%d:%d", major, minor)
+
+		f, err := os.Open(udevDataPath)
+		defer f.Close() // Closing this when it goes out of scope.
+		if err != nil {
+			// This path failed, try the fallback .udev style (non-systemd)
+			udevDataPath = fmt.Sprintf("/dev/.udev/db/block:%s", devName)
+			f, err2 := os.Open(udevDataPath)
```

Comment:
How about 
```suggestion
			f, err = os.Open(udevDataPath)
```
to avoid the ugly `err2`?